drivers/platform/stm32/stm32_spi.c

```diff
@@ -0,0 +1,232 @@
+	csip_extra.pull = GPIO_NOPULL;
+	csip_extra.mode = GPIO_MODE_OUTPUT_PP;
+	csip_extra.speed = GPIO_SPEED_FREQ_LOW;
+	csip.number = param->chip_select;
+	csip.extra = &csip_extra;
+	csip.platform_ops = &stm32_gpio_platform_ops;
+	ret = gpio_get(&sdesc->chip_select, &csip);
+	if (ret < 0)
+		goto error;
+
+	ret = gpio_direction_output(sdesc->chip_select, GPIO_HIGH);
+	if (ret < 0)
+		goto error;
+
+	/* automatically select prescaler based on max_speed_hz */
+	if (param->max_speed_hz != 0u && sinit->get_input_clock) {
+		input_clock = sinit->get_input_clock();
+		uint32_t div = input_clock / param->max_speed_hz;
+		uint32_t rem = input_clock % param->max_speed_hz;
+		uint32_t po2 = !(div & (div - 1)) && !rem;
+
+		// find the power of two just higher than div and
+		// store the exponent in prescaler_reg
+		while(div) {
+			prescaler_reg += 1;
+			div >>= 1u;
+		}
+
+		// this exponent - 1 is needed because of the way
+		// stm32 stores it into registers:
+		// reg = 0 eq. div = 2^1
+		// reg = 1 eq. div = 2^2 etc.
+		if (prescaler_reg)
+			prescaler_reg -= 1;
+
+		// this exponent - 1 is needed when initial div was
+		// precisely a power of two
+		if (prescaler_reg && po2)
+			prescaler_reg -= 1;
+
+		if (prescaler_reg < prescaler_min)
+			prescaler_reg = prescaler_min;
+
+		if (prescaler_reg > prescaler_max)
+			prescaler_reg = prescaler_max;
+	} else
+		prescaler_reg = prescaler_default;
+
+	sdesc->hspi.Instance = sinit->base;
+	sdesc->hspi.Init.Mode = SPI_MODE_MASTER;
+	sdesc->hspi.Init.Direction = SPI_DIRECTION_2LINES;
+	sdesc->hspi.Init.DataSize = SPI_DATASIZE_8BIT;
+	sdesc->hspi.Init.CLKPolarity = param->mode & SPI_CPOL ? SPI_POLARITY_HIGH :
+				       SPI_POLARITY_LOW;
+	sdesc->hspi.Init.CLKPhase = param->mode & SPI_CPHA ? SPI_PHASE_1EDGE :
+				    SPI_PHASE_2EDGE;
+	sdesc->hspi.Init.NSS = SPI_NSS_SOFT;
+	sdesc->hspi.Init.BaudRatePrescaler = prescaler_reg << SPI_CR1_BR_Pos;
+	sdesc->hspi.Init.FirstBit = param->bit_order ? SPI_FIRSTBIT_LSB :
+				    SPI_FIRSTBIT_MSB;
+	sdesc->hspi.Init.TIMode = SPI_TIMODE_DISABLE;
+	sdesc->hspi.Init.CRCCalculation = SPI_CRCCALCULATION_DISABLE;
+	sdesc->hspi.Init.CRCPolynomial = 10;
+	ret = HAL_SPI_Init(&sdesc->hspi);
+	if (ret != HAL_OK) {
+		ret = -EIO;
+		goto error;
+	}
+
+	/* copy settings to device descriptor */
+	(*desc)->max_speed_hz = param->max_speed_hz;
+	(*desc)->mode = param->mode;
+	(*desc)->bit_order = param->bit_order;
+	(*desc)->chip_select = param->chip_select;
+
+	return ret;
+error:
+	free(*desc);
+	free(sdesc);
+	return ret;
+}
+
+/**
+ * @brief Free the resources allocated by spi_init().
+ * @param desc - The SPI descriptor.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t stm32_spi_remove(struct spi_desc *desc)
+{
+	struct stm32_spi_desc *sdesc;
+
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	sdesc = desc->extra;
+	HAL_SPI_DeInit(&sdesc->hspi);
+	gpio_remove(sdesc->chip_select);
+	free(desc->extra);
+	free(desc);
+	return 0;
+}
```

Comment:
maybe `return SUCCESS`

Reply:
Unfortunately there is a name clash on stm32 platform with SUCCESS being already defined in STM32 HAL.
The only elegant way of overcoming this was use <errno.h> and bypass our "error.h" and use 0 instead of SUCCESS.